dags/common/sftp_service.py

```diff
@@ -55,9 +56,24 @@ def __exit__(self, exc_type, exc_value, tb):
             return False
         return True
 
-    def list_files(self):
+    def list_files(self, excluded_dirs=None):
         try:
-            return self.connection.listdir(self.dir)
+            file_names = []
+            self.connection.walktree(
+                self.dir,
+                lambda x: file_names.append(re.sub(self.dir + "/", "", x)),
+                lambda x: x,
+                lambda x: x,
+                recurse=True,
+            )
+            if excluded_dirs:
+                filtered_files = []
+                for excluded_dir in excluded_dirs:
+                    filtered_files += [
+                        file for file in file_names if excluded_dir not in file
+                    ]
+                return filtered_files
+            return file_names
```

Comment:
That's good, we can simplify it a lot by doing something like the following:

```python

    def list_files(self, exclude_directories=None):
        try:
            file_names = []
            self.connection.walktree(
                self.dir,
                lambda filename: append_file_if_not_in_excluded_directory(
                    filename, exclude_directories, file_names
                ),
                lambda directory: directory,
                lambda unknown: unknown,
                recurse=True,
            )
            return file_names
        except AttributeError:
            raise NotConnectedException
```

In utils.py

```python
def append_file_if_not_in_excluded_directory(filename, exclude_directories, list_of_files):
    if not exclude_directories or not (any(re.search(exclude, filename) for exclude in exclude_directories)) :
        list_of_files.append(filename)
```

And write a test about that case too in` integration/common/test_sftp.py`


```python
def test_list_files_with_exclude_directories():
    excluded_directories = ['JHEP']
    with SFTPService(dir="upload/springer/JHEP") as sftp:
        files = sftp.list_files(excluded_directories=excluded_directories)
        expected_files = []
        assert expected_files == files
```